src/api.ts

```diff
@@ -1,51 +1,189 @@
-import StoryblokClient from 'storyblok-js-client';
-import { session } from './session';
 import type { RegionCode } from './constants';
+import { getStoryblokUrl } from './utils/api-routes';
+import { delay, FetchError } from './utils/fetch';
 
-export interface ApiClientState {
-  region: RegionCode;
-  accessToken: string;
-  client: StoryblokClient | null;
+export interface ManagementApiClientOptions {
+  token?: string;
+  url?: string;
+  region?: RegionCode;
+  maxRetries?: number;
+  baseDelay?: number;
+  verbose?: boolean;
 }
 
-const state: ApiClientState = {
-  region: 'eu',
-  accessToken: '',
-  client: null,
-};
+export interface FetchOptions {
+  headers?: Record<string, string>;
+  method?: string;
+  body?: any;
+  maxRetries?: number;
+  baseDelay?: number;
+}
 
-export function apiClient() {
-  if (!state.client) {
-    createClient();
-  }
+export interface GetResponse<T> {
+  data: T;
+  attempt: number;
+}
 
-  function createClient() {
-    const userSession = session();
-    if (!userSession.state.isLoggedIn) {
-      throw new Error('User is not logged in');
+export interface MapiClient {
+  uuid: string;
+  get: <T>(path: string, fetchOptions?: FetchOptions) => Promise<GetResponse<T>>;
+  post: <T>(path: string, fetchOptions?: FetchOptions) => Promise<GetResponse<T>>;
+  put: <T>(path: string, fetchOptions?: FetchOptions) => Promise<GetResponse<T>>;
+  dispose: () => void;
+}
+
+let instance: MapiClient | null = null;
+
+const createMapiClient = (options: ManagementApiClientOptions): MapiClient => {
+  const baseHeaders = {
+    'Content-Type': 'application/json',
+    'Authorization': options.token,
+  };
+
+  const state = {
+    uuid: `mapi-client-${Math.random().toString(36).substring(2, 15)}`,
+    baseHeaders,
+    url: options.url || getStoryblokUrl(options.region),
+    maxRetries: options.maxRetries ?? 6,
+    baseDelay: options.baseDelay ?? 500,
+    freeze: false,
+  };
+
+  // Internal: isRateLimitOwner is true for the request that first hits 429 and is responsible for lifting the freeze
+  const request = async <T>(
+    path: string,
+    fetchOptions?: FetchOptions,
+    attempt: number = 0,
+    isRateLimitOwner: boolean = false,
+  ): Promise<GetResponse<T>> => {
+    // Only non-owner calls should wait for freeze
+    if (state.freeze && !isRateLimitOwner) {
+      if (options?.verbose) {
+        console.log(`⏳ ${path} - Waiting for rate limit to be resolved`);
+      }
+      await new Promise<void>((resolve) => {
+        const checkFreeze = setInterval(() => {
+          if (!state.freeze) {
+            clearInterval(checkFreeze);
+            resolve();
+          }
+        }, 50);
+      });
```

Comment:
A nicer alternative to this async loop would be to hold the value of state.freeze as a promise which can be awaited. This would avoid having to loop to check the state.

There's also a handy API for this. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/withResolvers